### PR TITLE
Redhat RSA FIPS patches

### PR DESCRIPTION
**Pull request by beldmit**
This PR brings a set of Red Hat RSA-specific FIPS-related patches upstream.

Original patches were implemented mostly by @neverpanic and @beldmit 

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Comment by beldmit** [comment]
**OTC: which patches of this subset are desired?**

I am going to fix the tests after we agree on which is acceptable.

File: providers/implementations/asymciphers/rsa_enc.c [hunk 1]
```diff
@@ -144,6 +155,18 @@ static int rsa_encrypt(void *vprsactx, unsigned char *out, size_t *outlen,
     if (!ossl_prov_is_running())
         return 0;
 
+# ifdef FIPS_MODULE
+    if (fips_padding_allowed(prsactx) == 0) {
+        ERR_raise(ERR_LIB_PROV, PROV_R_ILLEGAL_OR_UNSUPPORTED_PADDING_MODE);
+        return 0;
+    }
+
+    if (RSA_bits(prsactx->rsa) < OPENSSL_RSA_FIPS_MIN_MODULUS_BITS) {
```
Comment (slontis) [inline providers/implementations/asymciphers/rsa_enc.c:164]: There should already be a call to ossl_rsa_check_key(prsactx->libctx, vrsa, operation)

File: providers/implementations/asymciphers/rsa_enc.c [hunk 2]
```diff
@@ -206,6 +229,18 @@ static int rsa_decrypt(void *vprsactx, unsigned char *out, size_t *outlen,
...
```
Comment (slontis) [inline providers/implementations/asymciphers/rsa_enc.c:238]: same comment as above

File: providers/implementations/kem/rsa_kem.c [hunk 3]
```diff
@@ -265,6 +265,14 @@ static int rsasve_generate(PROV_RSA_CTX *prsactx,
             *secretlen = nlen;
         return 1;
     }
+
+#ifdef FIPS_MODULE
+    if (nlen < OPENSSL_RSA_FIPS_MIN_MODULUS_BITS/8) {
```
Comment (slontis) [inline providers/implementations/kem/rsa_kem.c:270]: rsakem_init() does a call to ossl_rsa_check_key(prsactx->libctx, vrsa, operation)

File: providers/implementations/signature/rsa_sig.c [hunk 4]
```diff
@@ -686,6 +686,19 @@ static int rsa_verify_recover(void *vprsactx,
 {
     PROV_RSA_CTX *prsactx = (PROV_RSA_CTX *)vprsactx;
     int ret;
+# ifdef FIPS_MODULE
+    size_t rsabits = RSA_bits(prsactx->rsa);
+
+    if (rsabits < 2048) {
+        if (rsabits != 1024
```
Comment (slontis) [inline providers/implementations/signature/rsa_sig.c:693]: I dont think this check is needed (or correct)

Comment (beldmit) [reply providers/implementations/signature/rsa_sig.c:693]: https://nvlpubs.nist.gov/nistpubs/SpecialPublications/NIST.SP.800-131Ar2.pdf section 3 table 2 and the text on page 10: "RSA: See FIPS 186-2 and FIPS 186-4, which include modulus lengths of 1024, 1280, 1536 and 1792 bits, may continue to be used for signature verification but not signature generation."

File: crypto/rsa/rsa_oaep.c [hunk 5]
```diff
@@ -73,9 +73,23 @@ int ossl_rsa_padding_add_PKCS1_OAEP_mgf1_ex(OSSL_LIB_CTX *libctx,
         return 0;
 #endif
     }
+
+#ifdef FIPS_MODULE
+    if (EVP_MD_is_a(md, "SHAKE-128") || EVP_MD_is_a(md, "SHAKE-256")) {
+        ERR_raise(ERR_LIB_RSA, RSA_R_DIGEST_NOT_ALLOWED);
+        return 0;
+    }
+#endif
```
Comment (paulidale) [inline crypto/rsa/rsa_oaep.c:82]: I'd suggest wrapping this into a utility function to save the repetition.

File: providers/implementations/signature/rsa_sig.c (hunk 4 above)
Comment (paulidale) [reply providers/implementations/signature/rsa_sig.c:693]: I'd read that differently to your lab.

It doesn't say: _which include **only** modulus lengths of_

Comment (beldmit) [reply providers/implementations/signature/rsa_sig.c:693]: Well... I believe that if the intention was not limit to specific lengths, the text would not say anything about particular modulus lengths.

Comment (paulidale) [reply providers/implementations/signature/rsa_sig.c:693]: We disagree :grin:

Comment (slontis) [reply providers/implementations/signature/rsa_sig.c:693]: My interpretation is the same as Pauli's.. 'Includes' is different from 'must be' one of..

File: providers/implementations/asymciphers/rsa_enc.c (hunk 1 above)
Comment (beldmit) [reply providers/implementations/asymciphers/rsa_enc.c:164]: Agreed.

File: providers/implementations/asymciphers/rsa_enc.c (hunk 2 above)
Comment (beldmit) [reply providers/implementations/asymciphers/rsa_enc.c:238]: No, for decryption we have a check for 1024 bits only: https://github.com/openssl/openssl/blob/master/providers/common/securitycheck.c#L39-L52

File: crypto/rsa/rsa_oaep.c (hunk 5 above)
Comment (beldmit) [reply crypto/rsa/rsa_oaep.c:82]: I agree. Do I correctly understand that this change is acceptable after refactoring?

File: providers/implementations/kem/rsa_kem.c (hunk 3 above)
Comment (beldmit) [reply providers/implementations/kem/rsa_kem.c:270]: Agreed

File: crypto/rsa/rsa_oaep.c (hunk 5 above)
Comment (paulidale) [reply crypto/rsa/rsa_oaep.c:82]: Best wait until after the OTC call.

**Comment by beldmit** [comment]
Current state of review:

> Check allowed lengths, forbid RSA KEX in TLS in FIPS mode

The length check is not accepted (@paulidale, @slontis), the TLS part is not reviewed.

> Forbid some RSA padding modes in FIPS provider

Looks like the length check for decrypt is for 1024 bits.
The padding checks are absent upstream.

> Remove X9.31 support from FIPS provider

No review yet

> Forbid SHAKE with RSA in FIPS mode

Presumably acceptable after moving the check to a separate function

> Check allowed lengths of RSA keys in RSA KEM implementation

The check in encapsulation is redundant, in decapsulation 1024 bits (as with decryption).

**Comment by beldmit** [comment]
Length checks (thanks @neverpanic )

The 1024 signature verification is explicitly permitted by SP 800-131Ar2 section 3, table 2. Key Agreement is handled in section 6, where table 5 disallows everything < 2048 bits. This also makes sense because Key Agreement is synchronous, while you may be verifying signatures made years ago.

So it's not just our understanding, it's explicitly specified in SP 800-131Ar2, which is the "Transition" document. That one is always a good source for specific key length limitations.

There's also this sentence which confirms there are no different limits for encryption and decryption: "In the case of key-transport keys (i.e., the keys used to encrypt other keys for transport), this document (SP 800-131A) applies to both the encryption and decryption of the transported keys."

File: crypto/rsa/rsa_oaep.c (hunk 5 above)
Comment (beldmit) [reply crypto/rsa/rsa_oaep.c:82]: Sure, now I'm trying to summarize what is **not** acceptable

**Comment by paulidale** [comment]
For a lot of these, we would have to guard the implementation using a flag as we've done for other FIPS restrictions that had to be introduced (EMS, DRBG non-truncated digests).

X9.31 would definitely fall into this category.  One could argue the SHAKE change could be done without since it's highly unlikely to be used anywhere (however the same was true of truncated digests with DRBGs).

**Comment by beldmit** [comment]
Quoting the OTC discussion:

> Updates to include various length checks and other misc updates.
Needs some updates for style and to remove things which are unacceptable.

> We would like all of the fixes eventually.

**Review by youppycestnet — APPROVED** [review]

File: providers/implementations/signature/rsa_sig.c (hunk 4 above)
Comment (jvdsn) [reply providers/implementations/signature/rsa_sig.c:693]: FIPS 140-3 IG C.F says

> Signature verification. When performing a FIPS 186-4 signature verification, a module may support all RSA modulus sizes no smaller than 2048, as well as a 1024-bit modulus. When performing a FIPS 186-2 signature verification, the module may support all modulus sizes shown in FIPS 186-2; that is, all numbers in the form of 1024+256*s, where s is a nonnegative integer. The use of RSA modulus sizes listed in this paragraph is approved for signature verification within their respective standard (FIPS 186-4 or FIPS 186-2).

also, for some background:

> Several NIST standards and the CMVP Guidance imply various restrictions on the sizes of the RSA moduli approved for use in the RSA signature generation and the RSA signature verification algorithms. FIPS 186-4approves the use of three modulus sizes: 1024, 2048, and 3072 bits. FIPS 186-2 defers to an ANSI standardX9.31, which permits the use of the RSA moduli sizes that can be represented as 1024 + 256*s bits, where s is a non-negative integer. SP 800-131A Rev2 approves for signature generation all RSA modulus sizes no smaller than 2048.

Does that change your interpretation?

**Comment by beldmit** [comment]
No reason to keep it open, upstream is implementing changes their own way

**Pull request closed without merging**